### PR TITLE
Added functionality to use Remote JUnit in older versions of OSGi

### PR DESCRIPTION
**Pull request by amitjoy**
1. Added Invariants
2. Code Cleanup

Signed-off-by: Amit Kumar Mondal <admin@amitinside.com>

### CODE REVIEW

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java [hunk 1]
```diff
@@ -36,47 +40,43 @@ public BundleBuilder x() {
...
-		}
+		Map<String, String> map = s.computeIfAbsent(outerkey, k -> new LinkedHashMap<>());
+		s.put(outerkey, map);
 		map.put(innerKey, value);
 	}
 
 	String join(String... packageNames) {
-		return Stream.of(packageNames)
-				.collect(Collectors.joining(","));
+		checkNonNullArgument(packageNames, "'packageNames' cannot be null");
```
Comment (bjhargrave) [inline biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java]: Use Objects.requireNonNull. But that is not necessary since I am sure Stream.of will NPE on a non-null arg.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java [hunk 2]
```diff
@@ -36,47 +40,43 @@ public BundleBuilder x() {
...
 
 	String prepare(String name, Map<String, Map<String, String>> domain) {
+		StringBuilder builder = new StringBuilder(name);
 		while (domain.containsKey(name)) {
-			name += "~";
+			builder.append("~");
 		}
-		domain.put(name, new LinkedHashMap<>());
-		return name;
+		domain.put(builder.toString(), new LinkedHashMap<>());
```
Comment (bjhargrave) [inline biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java]: Just call builder.toString() once. use a var to hold it for the 2 uses of the string. (Avoid rework).

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleSpecBuilder.java [hunk 3]
```diff
@@ -1,5 +1,8 @@
 package aQute.bnd.remote.junit;
 
+import static aQute.lib.preconditions.Preconditions.checkState;
```
Comment (bjhargrave) [inline biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleSpecBuilder.java]: Lets not use this new package. Let use standard Objects.requireNonNull for null checks and simple ifs for true/false checks.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleSpecBuilder.java [hunk 4]
```diff
@@ -205,55 +237,73 @@ public String fragmenthost() {
 	 * Extension options for the Fragment-Host
 	 */
 	enum Extension {
-		framework, bootclasspath
+		FRAMEWORK("framework"),
+		BOOTCLASSPATH("bootclasspath");
+
+		public final String value;
```
Comment (bjhargrave) [inline biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleSpecBuilder.java]: Don't make this public. Make it private, override toString and have it return value.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java [hunk 5]
```diff
@@ -311,45 +328,54 @@ public void report() throws InvalidSyntaxException {
...
 	 */
 	public void reportServices(Appendable out) throws InvalidSyntaxException {
+		checkNonNullArgument(out, "Output destination instance cannot be null");
 		try (Formatter f = new Formatter(out)) {
 			Stream.of(framework.getBundleContext()
-				.getAllServiceReferences(null, null))
-				.forEach(sref -> {
-					f.format("%s\n", sref);
-				});
+				.getServiceReferences((String) null, null))
```
Comment (bjhargrave) [inline biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: This is a semantic change not just a change for an older API. getAllServiceReferences does not care if the service type is not assignable to the caller's bundle. This change can return fewer services and thus may give incorrect results. getAllServiceReferences was added in R4.0, so this method should be used.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java [hunk 6]
```diff
@@ -311,45 +328,54 @@ public void report() throws InvalidSyntaxException {
...
-	 * @param timeoutInMs the time to wait
+	 * @param clazz the name of the service
+	 * @param timeout the time to wait
 	 * @return a service reference
 	 * @throws InterruptedException
+	 * @throws IllegalArgumentException if any of the specified arguments is
+	 *             {@code null}
 	 */
-	public <T> Optional<ServiceReference<T>> waitForServiceReference(Class<T> class1, long timeoutInMs)
+	public <T> Optional<ServiceReference<T>> waitForServiceReference(Class<T> clazz, Duration timeout)
```
Comment (bjhargrave) [inline biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: What is the point of the change to use Duration when you only get the millis out of it anyway? Seems an unnecessary change.

File: aQute.libg/src/aQute/lib/preconditions/Preconditions.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+	/**
+	 * Ensures that an object reference passed as a parameter to the calling
+	 * method is <b>not</b> {@code null}.
+	 *
+	 * @param obj an object reference
+	 * @param errorMessage the exception message to use if the check fails
+	 * @return the non-null reference that was validated
+	 * @throws IllegalArgumentException if {@code reference} is {@code null}
+	 */
+	public static <T> T checkNonNullArgument(T obj, String errorMessage) {
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply aQute.libg/src/aQute/lib/preconditions/Preconditions.java]: I agree. I did this since bnd usually throws `IllegalArgumentException` instead of NPE. Should I change it all of a sudden?

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java [hunk 8]
```diff
@@ -22,6 +25,7 @@
 	final List<Closeable>		closeables	= new ArrayList<>();
 
 	BundleBuilder(JUnitFramework ws) {
+		checkNonNullArgument(ws, "'ws' cannot be null");
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java]: I personally use Objects.requireNonNull everywhere but bnd throws `IllegalArgumentException` for invariants.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java (hunk 1 above)
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java]: True. It will definitely throw NPE. I just added the invariant to assert a proper message that the argument should not be null. I agree that I should use `Objects.requireNonNull` but didn't want to change the usual practice in bnd to throw `IllegalArgumentException` for invariants.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleSpecBuilder.java (hunk 3 above)
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleSpecBuilder.java]: Agreed.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleSpecBuilder.java (hunk 4 above)
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleSpecBuilder.java]: Good idea

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java (hunk 5 above)
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: ah you are right. will change it.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java (hunk 6 above)
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: From the provider perspective, I agree with you but from consumer perspective, it is more flexible to use duration in their code and pass it directly to the methods accepting `Duration` as argument. As a consumer, you have the flexibility to use any type of duration and don't have to consider about millis at all.

Comment (bjhargrave) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: With TimeUnit, you can easily convert anything to millis.

```
TimeUnit.SECONDS.toMillis(5);
TimeUnit.MINUTES.toMillis(1);
```

Comment (bjhargrave) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: That also nicely avoids creating objects since it all long arithmetic.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java [hunk 9]
```diff
@@ -214,19 +226,22 @@ public BundleContext getBundleContext() {
...
+	public <T> List<T> getServices(Class<T> clazz) {
+		requireNonNull(clazz, "Service Class instance cannot be null");
 		try {
-			Collection<ServiceReference<T>> refs = getBundleContext().getServiceReferences(class1, null);
-			List<T> result = new ArrayList<>();
-			for (ServiceReference<T> ref : refs) {
-				T service = getBundleContext().getService(ref);
-				if (service != null)
-					result.add(service);
+			ServiceReference[] refs = getBundleContext().getAllServiceReferences(clazz.getName(), null);
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: yes since we need to return all the services of the given type.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java [hunk 10]
```diff
@@ -459,12 +483,10 @@ private String bundleStateToString(int state) {
...
 	}
 
-	private List<? extends ServiceReference<?>> getReferences(Class<?> serviceClass, String target)
+	private List<? extends ServiceReference> getReferences(Class<?> serviceClass, String target)
 		throws InvalidSyntaxException {
-		List<? extends ServiceReference<?>> serviceReferences = new ArrayList<>(
-			getBundleContext().getServiceReferences(serviceClass, target));
-		Collections.sort(serviceReferences);
-		return serviceReferences;
+		ServiceReference[] refs = getBundleContext().getAllServiceReferences(serviceClass.getName(), target);
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: yes since we need to return all the services of the specified type.

File: cnf/ext/central.mvn [hunk 11]
```diff
@@ -37,6 +37,7 @@ commons-codec:commons-codec:1.10
 
 org.apache.felix:org.apache.felix.configadmin:1.2.4
 org.apache.felix:org.apache.felix.log:1.2.0
+org.apache.felix:org.apache.felix.framework:3.2.2
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply cnf/ext/central.mvn]: Since 3.2.2 is the implementation of OSGi 4.2 whereas the previously added Felix 4.2.1 is the implementation of OSGi 4.3

[the remark this answers is not in the document]
Comment (bjhargrave) [reply cnf/ext/central.mvn]: Perhaps, but we do not need to test on an old framework. We can test on a new framework since old bundles will run on new framework. So please use the newest Felix for tests.

**Comment by bjhargrave** [comment]
I released some fixes to the remote.junit tests which were failing on JDK 10. So you will need to rebase on master and fix any merge conflicts.

**Comment by amitjoy** [comment]
I also thought the same that it should work actually in newer versions of Felix too but the tests fail with the following stacktrace:

```
java.lang.NoSuchMethodError: org.osgi.framework.BundleEvent.<init>(ILorg/osgi/framework/Bundle;Lorg/osgi/framework/Bundle;)V
	at org.apache.felix.framework.Felix.fireBundleEvent(Felix.java:4584)
	at org.apache.felix.framework.Felix.installBundle(Felix.java:3199)
	at org.apache.felix.framework.BundleContextImpl.installBundle(BundleContextImpl.java:167)
	at aQute.bnd.remote.junit.JUnitFramework.testbundle(JUnitFramework.java:705)
	at aQute.bnd.remote.junit.JUnitFramework.start(JUnitFramework.java:664)
	at aQute.bnd.remote.junit.JUnitFrameworkBuilder.create(JUnitFrameworkBuilder.java:165)
	at aQute.bnd.remote.junit.HelloTest.testActivator(HelloTest.java:57)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:209)
```

File: cnf/ext/central.mvn (hunk 11 above)
[the remark this answers is not in the document]
Comment (amitjoy) [reply cnf/ext/central.mvn]: I also thought the same that it should work actually in newer versions of Felix too but the tests fail with the following stacktrace:

```
java.lang.NoSuchMethodError: org.osgi.framework.BundleEvent.<init>(ILorg/osgi/framework/Bundle;Lorg/osgi/framework/Bundle;)V
	at org.apache.felix.framework.Felix.fireBundleEvent(Felix.java:4584)
	at org.apache.felix.framework.Felix.installBundle(Felix.java:3199)
	at org.apache.felix.framework.BundleContextImpl.installBundle(BundleContextImpl.java:167)
	at aQute.bnd.remote.junit.JUnitFramework.testbundle(JUnitFramework.java:705)
	at aQute.bnd.remote.junit.JUnitFramework.start(JUnitFramework.java:664)
	at aQute.bnd.remote.junit.JUnitFrameworkBuilder.create(JUnitFrameworkBuilder.java:165)
	at aQute.bnd.remote.junit.HelloTest.testActivator(HelloTest.java:57)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:50)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:47)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.internal.runners.statements.RunAfters.evaluate(RunAfters.java:27)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:325)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:78)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:57)
	at org.junit.runners.ParentRunner$3.run(ParentRunner.java:290)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:71)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:288)
	at org.junit.runners.ParentRunner.access$000(ParentRunner.java:58)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:268)
	at org.junit.internal.runners.statements.RunBefores.evaluate(RunBefores.java:26)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:363)
	at org.eclipse.jdt.internal.junit4.runner.JUnit4TestReference.run(JUnit4TestReference.java:89)
	at org.eclipse.jdt.internal.junit.runner.TestExecution.run(TestExecution.java:41)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:541)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.runTests(RemoteTestRunner.java:763)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.run(RemoteTestRunner.java:463)
	at org.eclipse.jdt.internal.junit.runner.RemoteTestRunner.main(RemoteTestRunner.java:2
```

[the remark this answers is not in the document]
Comment (bjhargrave) [reply cnf/ext/central.mvn]: This must be because you have core-4.2 on the -runpath. The framework impl must supply the core api.

[the remark this answers is not in the document]
Comment (bjhargrave) [reply cnf/ext/central.mvn]: The core jar is only to be used for compilation. It should never be on the classpath when running a framework.

[the remark this answers is not in the document]
Comment (amitjoy) [reply cnf/ext/central.mvn]: Agreed but how did the the API jar end up in `-runpath`? It is only specified in `-buildpath`.

[the remark this answers is not in the document]
Comment (bjhargrave) [reply cnf/ext/central.mvn]: My guess is https://github.com/bndtools/bndtools/issues/1922 which is fixed in the current 4.2 bndtools builds. Are you using the latest 4.2 bndtools? You may find the test passes from the gradle build since it did not have the above issue.

[the remark this answers is not in the document]
Comment (amitjoy) [reply cnf/ext/central.mvn]: I am using the recent snapshot version but the gradle build is still failing. I am not sure if it is the transitive classpath calculation problem. The `System.getProperty("java.class.path")` also returns the following: 

```
System.getProperty("java.class.path")
	 (java.lang.String) /Users/amitmondal/git/bnd/biz.aQute.bnd.remote.junit/bin:/Users/amitmondal/git/bnd/biz.aQute.bnd.remote.junit/bin_test:/Users/amitmondal/.m2/repository/org/osgi/osgi.annotation/7.0.0/osgi.annotation-7.0.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.core/4.2.0/org.osgi.core-4.2.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.service.component/1.3.0/org.osgi.service.component-1.3.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.service.component.annotations/1.3.0/org.osgi.service.component.annotations-1.3.0.jar:/Users/amitmondal/git/bnd/aQute.libg/bin:/Users/amitmondal/git/bnd/aQute.libg/bin_test:/Users/amitmondal/.m2/repository/org/slf4j/slf4j-api/1.7.25/slf4j-api-1.7.25.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.util.function/1.1.0/org.osgi.util.function-1.1.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.util.promise/1.1.1-SNAPSHOT/org.osgi.util.promise-1.1.1-SNAPSHOT.jar:/Users/amitmondal/.m2/repository/org/apache/servicemix/bundles/org.apache.servicemix.bundles.junit/4.12_1/org.apache.servicemix.bundles.junit-4.12_1.jar:/Users/amitmondal/.m2/repository/org/assertj/assertj-core/3.10.0/assertj-core-3.10.0.jar:/Users/amitmondal/.m2/repository/org/osgi/osgi.core/6.0.0/osgi.core-6.0.0.jar:/Users/amitmondal/.m2/repository/org/slf4j/slf4j-simple/1.7.25/slf4j-simple-1.7.25.jar:/Users/amitmondal/git/bnd/biz.aQute.bndlib/bin:/Users/amitmondal/git/bnd/biz.aQute.bndlib/bin_test:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.namespace.contract/1.0.0/org.osgi.namespace.contract-1.0.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.namespace.extender/1.0.1/org.osgi.namespace.extender-1.0.1.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.namespace.implementation/1.0.0/org.osgi.namespace.implementation-1.0.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.namespace.service/1.0.0/org.osgi.namespace.service-1.0.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.service.log/1.3.0/org.osgi.service.log-1.3.0.jar:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.service.repository/1.1.0/org.osgi.service.repository-1.1.0.jar:/Users/amitmondal/git/bnd/biz.aQute.bnd.annotation/bin:/Users/amitmondal/git/bnd/biz.aQute.bnd.annotation/bin_test:/Users/amitmondal/git/bnd/biz.aQute.bndlib/generated/biz.aQute.bndlib.jar:/Users/amitmondal/git/bnd/biz.aQute.repository/bin:/Users/amitmondal/git/bnd/biz.aQute.repository/bin_test:/Users/amitmondal/.m2/repository/org/osgi/org.osgi.service.coordinator/1.0.2/org.osgi.service.coordinator-1.0.2.jar:/Users/amitmondal/.m2/repository/org/tukaani/xz/1.5/xz-1.5.jar:/Users/amitmondal/.m2/repository/org/mockito/mockito-core/2.23.0/mockito-core-2.23.0.jar:/Users/amitmondal/.m2/repository/net/bytebuddy/byte-buddy/1.9.0/byte-buddy-1.9.0.jar:/Users/amitmondal/.m2/repository/org/objenesis/objenesis/2.6/objenesis-2.6.jar:/Users/amitmondal/.m2/repository/org/apache/geronimo/specs/geronimo-servlet_2.5_spec/1.2/geronimo-servlet_2.5_spec-1.2.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-continuation/7.6.3.v20120416/jetty-continuation-7.6.3.v20120416.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-http/7.6.3.v20120416/jetty-http-7.6.3.v20120416.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-io/7.6.3.v20120416/jetty-io-7.6.3.v20120416.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-security/7.6.3.v20120416/jetty-security-7.6.3.v20120416.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-server/7.6.3.v20120416/jetty-server-7.6.3.v20120416.jar:/Users/amitmondal/.m2/repository/org/eclipse/jetty/jetty-util/7.6.3.v20120416/jetty-util-7.6.3.v20120416.jar:/Users/amitmondal/git/bnd/biz.aQute.http.testservers/bin:/Users/amitmondal/git/bnd/biz.aQute.http.testservers/bin_test:/Users/amitmondal/.m2/repository/org/nanohttpd/nanohttpd/2.2.0/nanohttpd-2.2.0.jar:/Users/amitmondal/.m2/repository/org/bouncycastle/bcpkix-jdk15on/1.54/bcpkix-jdk15on-1.54.jar:/Users/amitmondal/.m2/repository/org/bouncycastle/bcprov-jdk15on/1.54/bcprov-jdk15on-1.54.jar:/Users/amitmondal/.m2/repository/org/littleshoot/littleproxy/1.1.2/littleproxy-1.1.2.jar:/Users/amitmondal/.m2/repository/com/google/guava/guava/19.0/guava-19.0.jar:/Users/amitmondal/.m2/repository/org/apache/commons/commons-lang3/3.4/commons-lang3-3.4.jar:/Users/amitmondal/.m2/repository/commons-codec/commons-codec/1.10/commons-codec-1.10.jar:/Users/amitmondal/git/bnd/biz.aQute.http.testservers/generated/biz.aQute.http.testservers.jar:/Users/amitmondal/git/bnd/biz.aQute.repository/generated/biz.aQute.repository.jar:/Users/amitmondal/java-latest/Eclipse.app/Contents/Eclipse/configuration/org.eclipse.osgi/225/0/.cp:/Users/amitmondal/java-latest/Eclipse.app/Contents/Eclipse/configuration/org.eclipse.osgi/224/0/.cp:/Users/amitmondal/java-latest/Eclipse.app/Contents/Eclipse/configuration/org.eclipse.osgi/227/0/.cp/lib/javaagent-shaded.jar
```

It comprises both OSGi Core 4.2 and OSGi Core 6.0 which seems really strange.

[the remark this answers is not in the document]
Comment (bjhargrave) [reply cnf/ext/central.mvn]: This may be a defect in the design of JunitFramework in that it is too connected to the build environment of the test project. I don't have any real experience using JunitFramework but it seems to me that there ought to be a disconnect between the class path of the framework being tested and the class path of the test case. It seems that the test case class path bleeds into the framework being tested which mean you have to have an exact match of the core version in test case class path and the framework being tested. While this could be matched (e.g. use a 4.2 framework version here), it is entirely permissible for framework implementations to customize the implementation of the core classes and so the framework must be able to use its own core classes rather than the vanilla ones used to build the test case. 

I think you need to take this up with @pkriens.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java]: Actually you can pass `null` in lieu of varargs.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java [hunk 12]
```diff
@@ -214,19 +227,20 @@ public BundleContext getBundleContext() {
...
 		try {
-			Collection<ServiceReference<T>> refs = getBundleContext().getServiceReferences(class1, null);
-			List<T> result = new ArrayList<>();
-			for (ServiceReference<T> ref : refs) {
-				T service = getBundleContext().getService(ref);
-				if (service != null)
-					result.add(service);
-			}
-			return result;
+			ServiceReference<?>[] refs = getBundleContext().getServiceReferences(clazz.getName(), null);
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: True. It's hard to debug in streams.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/Service.java [hunk 13]
```diff
@@ -21,10 +21,10 @@
 	/**
 	 * Target filter
 	 */
-	String target() default "";
+	String filter() default "";
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/Service.java]: Agreed. It could therefore align with the `@Reference` type.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/Service.java [hunk 14]
```diff
@@ -21,10 +21,10 @@
...
 
 	/**
 	 * The service type
 	 */
-	Class<?> service() default Object.class;
+	Class<?> type() default Object.class;
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/Service.java]: Agreed. It could therefore align with the `@Reference` type.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java [hunk 15]
```diff
@@ -36,47 +39,47 @@ public BundleBuilder x() {
...
+		Map<String, String> map = s.computeIfAbsent(outerkey, k -> new LinkedHashMap<>());
+		s.put(outerkey, map);
 		map.put(innerKey, value);
 	}
 
 	String join(String... packageNames) {
-		return Stream.of(packageNames)
-				.collect(Collectors.joining(","));
+		requireNonNull(packageNames, "'packageNames' cannot be null");
+		return Stream.of(packageNames).collect(Collectors.joining(","));
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/BundleBuilder.java]: Should the streams be formatted differently in bnd?

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java [hunk 16]
```diff
@@ -311,73 +327,88 @@ public void report() throws InvalidSyntaxException {
...
+		LocalDateTime endTime = currentTime.plus(timeout);
+
+		while (endTime.isAfter(currentTime)) {
+			@SuppressWarnings("unchecked")
+			ServiceReference serviceReference = getBundleContext().getServiceReference(clazz.getName());
 			if (serviceReference != null) {
 				return Optional.of(serviceReference);
 			}
-			Thread.sleep(50);
+			TimeUnit.MILLISECONDS.sleep(50);
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: It was just a matter of taste for readability.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java [hunk 17]
```diff
@@ -311,73 +327,88 @@ public void report() throws InvalidSyntaxException {
...
+	public <T> Optional<T> waitForService(Class<T> clazz, Duration timeout) throws InterruptedException {
+		requireNonNull(clazz, "Service class instance cannot be null");
+		requireNonNull(timeout, "Timeout cannot be null");
+
 		try {
-			Optional<ServiceReference<T>> ref = waitForServiceReference(class1, timeoutInMs);
+			Optional<ServiceReference> ref = waitForServiceReference(clazz, timeout);
 			BundleContext context = getBundleContext();
-			return ref.map(context::getService);
+			return ref.map(context::getService).filter(clazz::isInstance).map(clazz::cast);
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: I agree, it will always be the instance of the type provided.

File: biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java [hunk 18]
```diff
@@ -494,50 +516,40 @@ private boolean isMultiple(Type type) {
...
 		return false;
 	}
 
 	private boolean isParameterizedType(Type to, Class<?> clazz) {
-		if (to instanceof ParameterizedType) {
-			if (((ParameterizedType) to).getRawType() == clazz)
-				return true;
-		}
-		return false;
+		return to instanceof ParameterizedType && ((ParameterizedType) to).getRawType() == clazz;
```
[the remark this answers is not in the document]
Comment (amitjoy) [reply biz.aQute.bnd.remote.junit/src/aQute/bnd/remote/junit/JUnitFramework.java]: I completely agree that the PR became effectively big due to cleanups as I wanted my static analyzers not to complain about the small things in the code. That's how the changes got accumulated. Next time I will definitely concentrate only on the new features instead of making the static analyzer happy 😃
